docs/design/extending-api.md

```diff
@@ -0,0 +1,255 @@
+
+
+## Storage
+
+In order to store custom user data in a versioned fashion inside of etcd, we need to also introduce a
+`Codec` compatible object for persistent storage in etcd.  This object is `ThirdPartyObjectData` and it contains:
+   * Standard API Metadata
+   * `Data`: The raw JSON data for this custom object.
+
+### Storage key specification
+
+Each custom object stored by the API server needs a custom key in storage, this is described below:
+
+#### Definitions
+
+   * `resource-namespace` : the namespace of the particular resource that is being stored
+   * `resource-name`: the name of the particular resource being stored
+   * `third-party-resource-namespace`: the namespace of the `ThirdPartyResource` resource that represents the type for the specific instance being stored.
+   * `third-party-resource-name`: the name of the `ThirdPartyResource` resource that represents the type for the specific instance being stored.
+
+#### Key
+
+Given the definitions above, the key for a specific third-party object is:
+
+```
+${standard-k8s-prefix}/third-party-objects/${third-party-resource-namespace}/${third-party-resource-name}/${resource-namespace}/${resource-name}
+```
+
+Thus, listing a third-party resource can be achieved by listing the directory:
+
+```
+${standard-k8s-prefix}/third-party-objects/${third-party-resource-namespace}/${third-party-resource-name}/${resource-namespace}/
+```
+
+
+## Object details
+
+```go
+// A ThirdPartyObject is a generic representation of a resource, it is used by add-ons and plugins to add new resource
+// types to the API.  It consists of one or more Versions of the api.
+type ThirdPartyObject struct {
+     TypeMeta   `json:",inline"`
+     ObjectMeta `json:"metadata,omitempty" description:"standard object metadata"`
+
+     Description string `json:"description,omitempty" description:"The description of this object"`
+
+     Versions []Version `json:"versions,omitempty" description:"The versions for this third party object"`
+}
+
+type ThirdPartyObjectList struct {
+     TypeMeta `json:",inline"`
+     ListMeta `json:"metadata,omitempty" description:"standard list metadata; see http://docs.k8s.io/api-conventions.md#metadata"`
+
+     Items []ThirdPartyObject `json:"items" description:"items is a list of schema objects"`
+}
+
+// A Version represents a single concrete version of an object model.
+type Version struct {
+     Name     string `json:"name,omitempty" description:"name of this version (e.g. 'v1')"`
+     APIGroup string `json:"apiGroup,omitempty" description:"The API group to add this object into, default 'experimental'"`
+}
+
+// An internal object, used for versioned storage in etcd.  Not exposed to the end user.
+type ThirdPartyObjectData struct {
+     TypeMeta   `json:",inline"`
+     ObjectMeta `json:"metadata,omitempty" description:"standard object metadata"`
+
+     Data string `json:"name,omitempty" description:"the raw JSON data for this data"`
```

Comment:
You changed this to `[]byte` in the code.

Just delete the object details. You've already demonstrated that they won't be kept up to date.